src/algebra/ring/boolean_ring.lean

```diff
@@ -0,0 +1,144 @@
+/-
+Copyright (c) 2021 Bryan Gin-ge Chen. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bryan Gin-ge Chen
+-/
+
+import tactic.ring
+import tactic.abel
+
+/-!
+# Boolean rings
+
+A Boolean ring is a ring where multiplication is idempotent. They are equivalent to Boolean
+algebras.
+
+## Main declarations
+
+* `boolean_ring`: a typeclass for rings where multiplication is idempotent.
+* `boolean_ring.to_boolean_algebra`: every Boolean ring is a Boolean algebra; this definition and
+  the `sup` and `inf` notations for `boolean_ring` are localized as instances in the
+  `boolean_algebra_of_boolean_ring` locale.
+
+## Tags
+
+boolean ring, boolean algebra
+
+-/
+
+/-- A Boolean ring is a ring where multiplication is idempotent. -/
+class boolean_ring α extends ring α :=
+(mul_idem : ∀ a : α, a * a = a)
```

Comment:
I don't have a strong opinion, but should this maybe be called `mul_self`?
Whatever you choose, it would be good to repeat this fact in the root namespace (and make it a simp lemma?)

Reply:
Ah, I was following `inf_idem` and `sup_idem`, but `mul_self` would probably be easier to guess in the context of rings.

Reply:
We can of course just have both